pkg/server/apiextensions.go

```diff
@@ -31,137 +31,232 @@ import (
 	apiextensionslisters "k8s.io/apiextensions-apiserver/pkg/client/listers/apiextensions/v1"
 	"k8s.io/apimachinery/pkg/api/equality"
 	apierrors "k8s.io/apimachinery/pkg/api/errors"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
 	"k8s.io/apimachinery/pkg/labels"
 	"k8s.io/apimachinery/pkg/runtime/schema"
+	"k8s.io/apimachinery/pkg/util/sets"
 	genericapirequest "k8s.io/apiserver/pkg/endpoints/request"
 	"k8s.io/client-go/tools/clusters"
 	"k8s.io/klog/v2"
 
+	apisv1alpha1 "github.com/kcp-dev/kcp/pkg/apis/apis/v1alpha1"
+	tenancyv1alpha1 "github.com/kcp-dev/kcp/pkg/apis/tenancy/v1alpha1"
 	"github.com/kcp-dev/kcp/pkg/apis/tenancy/v1alpha1/helper"
+	kcpclientset "github.com/kcp-dev/kcp/pkg/client/clientset/versioned"
+	apislisters "github.com/kcp-dev/kcp/pkg/client/listers/apis/v1alpha1"
 	tenancylisters "github.com/kcp-dev/kcp/pkg/client/listers/tenancy/v1alpha1"
 )
 
-// inheritanceCRDLister is a CRD lister that add support for ClusterWorkspace API inheritance.
-type inheritanceCRDLister struct {
-	crdLister       apiextensionslisters.CustomResourceDefinitionLister
-	workspaceLister tenancylisters.ClusterWorkspaceLister
+// SystemCRDLogicalCluster is the logical cluster we install system CRDs into for now. These are needed
+// to start wildcard informers until a "real" workspace gets them installed.
+const SystemCRDLogicalCluster = helper.LocalSystemClusterPrefix + "system-crds"
+
+type systemCRDProvider struct {
+	rootCRDs      sets.String
+	orgCRDs       sets.String
+	universalCRDs sets.String
+
+	getClusterWorkspace func(ctx context.Context, key string) (*tenancyv1alpha1.ClusterWorkspace, error)
+	getCRD              func(key string) (*apiextensionsv1.CustomResourceDefinition, error)
+}
+
+// NewSystemCRDProvider returns CRDs for certain cluster workspace types and the root workspace.
+// TODO(sttts): This must be replaced by some non-hardcoded mechanism in the (near) future, probably by
```

Comment:
It is for all CRDs (and APIBindings). The deserializer in the request chain prunes. If you use a too small schema (i.e. you have different CRDs and you choose one in the request handling of wildcard requests with a schema that lacks fields), you basically risk data-loss because the client gets incomplete objects.